### ISSUE TITLE
LicenseHeaderManager throws "The parameter is incorrect. (Exception from HRESULT: 0x80070057 (E_INVALIDARG))." when trying to add the LicenseHeader

### ISSUE THREAD

**Issue by FlorianDecker** [rubicon-oss/LicenseHeaderManager#68]
Reported in #57

**Comment by Convery** [comment]
For VS2015 Community on Win10 Pro 64-bit: 
v1.7.2 - Works
v1.7.3 - Error
v2.0.0 - Error

So either VS wants extra params for unsigned packages or v1.7.3 introduced something strange. Would anyone be able to package 1.7.2 for VS2017 just to test?

**Comment by FlorianDecker** [comment]
Hi!

Attached is a unsigned 1.7.2.1 Version with VS2017 compatibility. You will probably get a lot of "Settings update to 1.7.2.1 didnt work" errors, as they appear when downgrading LHM, an error which was only fixed in 1.7.3.

[LicenseHeaderManager_1.7.2.1_unsigned.zip](https://github.com/rubicon-oss/LicenseHeaderManager/files/875457/LicenseHeaderManager_1.7.2.1_unsigned.zip)

But i think that your error is probably due to some bug introduced with 1.7.3. Does the error appear with, as example, a fresh C# Class library project or only in your specific project? If it is your specific project, could you share either the project or the type of the project?

**Comment by Convery** [comment]
The settings update errors aside (just ignored for each popup), the v1.7.2.1 plugin worked flawlessly. Tested in an existing C++ project, and a new C# project. So it's indeed something introduced in 1.7.3, sadly I can't provide any useful debugging for it.

**Comment by mechacrash** [comment]
I can confirm this works on my installation too (cleared local settings to stop the perpetually error messages).

**Comment by FlorianDecker** [comment]
Ok, well we tried :)

If you have the time to assist us to catch that bug, would you be able to test various LHM Versions to narrow down where the bug appeared? Optimally would be of course, if you would be able to debug it (you would have to download the repository, build it, install the resulting .vsix to your Visual Studio and then press play.)

If you want to test my built LHM versions, I would just quick and dirty compile the LHM at certain commits and attach it in this thread. As it would be between 1.7.2 and 1.7.3 release, you would have to test it with VS2015.

Optimal workflow would be:

1. Uninstall current LHM
2. Close Visual Studio
3. Install new LHM with double click on the supplied .vsix
4. Open Visual Studio on an C# Project
5. Right Click .cs File -> Add License Header

I am going to start with the .vsix where we think is the highest possibility for the error:

First: 1.7.3, but compiled on my Computer and not the computer of MichaelKetting (to confirm the error is not introduced while building)
[LicenseHeaderManager_1_7_3_unsigned_FD.zip](https://github.com/rubicon-oss/LicenseHeaderManager/files/875790/LicenseHeaderManager_1_7_3_unsigned_FD.zip)

Second, Commit "f6bd6f": From relatively the middle of the 1.7.3, directlry after the fix for the trace.assert message (44). Contains fixes for 44, 42, 37, 63 and 56. Is not containing fixes for 47, 59, 62, 64 and 65.
[LicenseHeaderManager_f6bd6f_unsigned.zip](https://github.com/rubicon-oss/LicenseHeaderManager/files/876176/LicenseHeaderManager_f6bd6f_unsigned.zip)

Thanks for your help to squash that bug. We really need to introduce logging and improve the exception behaviour.

@mechacrash hmhh, thats interesting. Nice that it is working, but i have no idea why ;)

**Comment by mirdus** [comment]
Hi, I've tracked down the issue and it happens in the GetPath method in the ProjectItemParentFinder. I built a fix which seems to work on my side:

        private static string GetPath(ProjectItem projectItem)
        {
            if (projectItem.Properties == null)
                return String.Empty;

            var fullPathProperty = projectItem.Properties.Cast<Property>().FirstOrDefault(property => property.Name == "FullPath");
            if (fullPathProperty != null && fullPathProperty.Value != null)
                return fullPathProperty.Value.ToString();
            else
                return String.Empty;
        }

Rgds,

M.

**Comment by FlorianDecker** [comment]
Awesome work!

@mechacrash @Convery could you verify that version 2.0.0 with the fix supplied by mirdus fixes your problem too? I attached a compiled version.

[LicenseHeaderManager_2.0.0_WithFix_unsigned.zip](https://github.com/rubicon-oss/LicenseHeaderManager/files/934309/LicenseHeaderManager_2.0.0_WithFix_unsigned.zip)

**Comment by Convery** [comment]
Yep, the .zip works on my 2017 community edition.

**Comment by Convery** [comment]
Cool, VS2017 updated and now we're back to `(Exception from HRESULT: 0x80070057 (E_INVALIDARG))`

**Comment by FlorianDecker** [comment]
Maybe VS2017 redownloaded the online Version? Try reinstalling the version posted here. 
We currently have not published the fix in an official Version.

I tried again to reproduce it and still can't :/. Also the bug mirdus fixed does not seem to appear anywhere else (where a LicenseHeader would be added).

**Comment by coinoperated0428** [comment]
Has this been resolved? I get the error on C/C++ files but only when I do individual files.  Updating the complete project works. I'm on 2015 Enterprise Update 3 version 2.0.0 of LHM. Not see the error on C# code.

**Comment by learn-more** [comment]
Visual studio 2015 update 3, applying to a single c++ file happens for me as well. (LHM 2.0.0)
```
---------------------------
Microsoft Visual Studio
---------------------------
The parameter is incorrect. (Exception from HRESULT: 0x80070057 (E_INVALIDARG))
---------------------------
OK   
---------------------------

```

**Comment by learn-more** [comment]
Here is the callstack (which already triggers, first chance) upon right clicking a .c file in the solution explorer:

```
>	LicenseHeaderManager.dll!LicenseHeaderManager.Utils.ProjectItemInspection.IsLink(EnvDTE.ProjectItem projectItem)	Unknown
 	LicenseHeaderManager.dll!LicenseHeaderManager.Headers.LicenseHeaderReplacer.TryCreateDocument(EnvDTE.ProjectItem item, out LicenseHeaderManager.Headers.Document document, out bool wasOpen, System.Collections.Generic.IDictionary<string, string[]> headers)	Unknown
 	LicenseHeaderManager.dll!LicenseHeaderManager.LicenseHeadersPackage.ShouldBeVisible(EnvDTE.ProjectItem item)	Unknown
 	LicenseHeaderManager.dll!LicenseHeaderManager.LicenseHeadersPackage.QueryProjectItemCommandStatus(object sender, System.EventArgs e)	Unknown
 	Microsoft.VisualStudio.Shell.10.0.dll!Microsoft.VisualStudio.Shell.OleMenuCommand.OleStatus.get()	Unknown
 	Microsoft.VisualStudio.Shell.10.0.dll!Microsoft.VisualStudio.Shell.OleMenuCommandService.Microsoft.VisualStudio.OLE.Interop.IOleCommandTarget.QueryStatus(ref System.Guid guidGroup, uint nCmdId, Microsoft.VisualStudio.OLE.Interop.OLECMD[] oleCmd, System.IntPtr oleText)	Unknown
 	Microsoft.VisualStudio.Shell.10.0.dll!Microsoft.VisualStudio.Shell.Package.Microsoft.VisualStudio.OLE.Interop.IOleCommandTarget.QueryStatus(ref System.Guid guidGroup, uint nCmdId, Microsoft.VisualStudio.OLE.Interop.OLECMD[] oleCmd, System.IntPtr oleText)	Unknown
 	[Native to Managed Transition]	
 	[Managed to Native Transition]	
 	Microsoft.VisualStudio.Shell.14.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.HierarchyUtilities.ExecHierParentChain.AnonymousMethod__1()	Unknown
 	Microsoft.VisualStudio.Shell.14.0.dll!Microsoft.VisualStudio.ErrorHandler.CallWithCOMConvention(System.Func<int> method, bool reportError, bool setShellErrorInfo)	Unknown
 	Microsoft.VisualStudio.Shell.14.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.HierarchyUtilities.ExecHierParentChain(Microsoft.VisualStudio.Shell.IVsHierarchyItemManager manager, Microsoft.VisualStudio.Shell.Interop.IVsUIHierarchy lpUIHCmd, Microsoft.VisualStudio.Shell.Interop.IVsUIHierarchy lpUIHCurrent, uint itemidCurrent, ref System.Guid pguidCmdGroupRef, uint nCmdID, uint nCmdexecopt, System.IntPtr pvaIn, System.IntPtr pvaOut)	Unknown
 	Microsoft.VisualStudio.Shell.TreeNavigation.HierarchyProvider.dll!Microsoft.VisualStudio.PlatformUI.HierarchyItem.HierarchyContextMenuController.ShowContextMenu(System.Collections.Generic.IEnumerable<object> items, System.Windows.Point location)	Unknown
 	Microsoft.VisualStudio.Shell.14.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.ContextMenuController.ShowContextMenu(System.Collections.Generic.IEnumerable<Microsoft.Internal.VisualStudio.PlatformUI.IContextMenuPattern> items, System.Windows.Point location, System.Func<Microsoft.Internal.VisualStudio.PlatformUI.IContextMenuPattern, Microsoft.Internal.VisualStudio.PlatformUI.IContextMenuController> getController)	Unknown
 	Microsoft.VisualStudio.Shell.14.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.ContextMenuController.ShowContextMenu(System.Collections.Generic.IEnumerable<Microsoft.Internal.VisualStudio.PlatformUI.IContextMenuPattern> items, System.Windows.Point location)	Unknown
 	Microsoft.VisualStudio.Shell.14.0.dll!Microsoft.Internal.VisualStudio.PlatformUI.PivotTreeView.OnContextMenuOpening(System.Windows.Controls.ContextMenuEventArgs e)	Unknown
 	PresentationFramework.dll!System.Windows.FrameworkElement.OnContextMenuOpeningThunk(object sender, System.Windows.Controls.ContextMenuEventArgs e)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.ContextMenuEventArgs.InvokeEventHandler(System.Delegate genericHandler, object genericTarget)	Unknown
 	PresentationCore.dll!System.Windows.RoutedEventArgs.InvokeHandler(System.Delegate handler, object target)	Unknown
 	PresentationCore.dll!System.Windows.RoutedEventHandlerInfo.InvokeHandler(object target, System.Windows.RoutedEventArgs routedEventArgs)	Unknown
 	PresentationCore.dll!System.Windows.EventRoute.InvokeHandlersImpl(object source, System.Windows.RoutedEventArgs args, bool reRaised)	Unknown
 	PresentationCore.dll!System.Windows.UIElement.RaiseEventImpl(System.Windows.DependencyObject sender, System.Windows.RoutedEventArgs args)	Unknown
 	PresentationCore.dll!System.Windows.ContentElement.RaiseTrustedEvent(System.Windows.RoutedEventArgs args)	Unknown
 	PresentationCore.dll!System.Windows.ContentElement.RaiseEvent(System.Windows.RoutedEventArgs args, bool trusted)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.PopupControlService.RaiseContextMenuOpeningEvent(System.Windows.IInputElement source, double x, double y, bool userInitiated)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.PopupControlService.ProcessMouseUp(object sender, System.Windows.Input.MouseButtonEventArgs e)	Unknown
 	PresentationFramework.dll!System.Windows.Controls.PopupControlService.OnPostProcessInput(object sender, System.Windows.Input.ProcessInputEventArgs e)	Unknown
 	PresentationCore.dll!System.Windows.Input.InputManager.RaiseProcessInputEventHandlers(System.Windows.Input.ProcessInputEventHandler postProcessInput, System.Windows.Input.ProcessInputEventArgs processInputEventArgs)	Unknown
 	PresentationCore.dll!System.Windows.Input.InputManager.ProcessStagingArea()	Unknown
 	PresentationCore.dll!System.Windows.Input.InputManager.ProcessInput(System.Windows.Input.InputEventArgs input)	Unknown
 	PresentationCore.dll!System.Windows.Input.InputProviderSite.ReportInput(System.Windows.Input.InputReport inputReport)	Unknown
 	PresentationCore.dll!System.Windows.Interop.HwndMouseInputProvider.ReportInput(System.IntPtr hwnd, System.Windows.Input.InputMode mode, int timestamp, System.Windows.Input.RawMouseActions actions, int x, int y, int wheel)	Unknown
 	PresentationCore.dll!System.Windows.Interop.HwndMouseInputProvider.FilterMessage(System.IntPtr hwnd, MS.Internal.Interop.WindowMessage msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled)	Unknown
 	PresentationCore.dll!System.Windows.Interop.HwndSource.InputFilterMessage(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled)	Unknown
 	WindowsBase.dll!MS.Win32.HwndWrapper.WndProc(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam, ref bool handled)	Unknown
 	WindowsBase.dll!MS.Win32.HwndSubclass.DispatcherCallbackOperation(object o)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.InternalRealCall(System.Delegate callback, object args, int numArgs)	Unknown
 	WindowsBase.dll!System.Windows.Threading.ExceptionWrapper.TryCatchWhen(object source, System.Delegate callback, object args, int numArgs, System.Delegate catchHandler)	Unknown
 	WindowsBase.dll!System.Windows.Threading.Dispatcher.LegacyInvokeImpl(System.Windows.Threading.DispatcherPriority priority, System.TimeSpan timeout, System.Delegate method, object args, int numArgs)	Unknown
 	WindowsBase.dll!MS.Win32.HwndSubclass.SubclassWndProc(System.IntPtr hwnd, int msg, System.IntPtr wParam, System.IntPtr lParam)	Unknown

```

**Comment by FlorianDecker** [comment]
@learn-more thanks for the bug report! I think i fixed it, could you try this version to check if the fix worked?

@coinoperated0428 maybe you suffer from the same issue as @learn-more ? You could also try this version:

[LicenseHeaderManager_2.0.1_with_IsLinkFix_unsigned.zip](https://github.com/rubicon-oss/LicenseHeaderManager/files/1281280/LicenseHeaderManager_2.0.1_with_IsLinkFix_unsigned.zip)


Currently we havent released this fix to the Visual Studio Gallery due to time constraints. Thanks for your support!

**Comment by learn-more** [comment]
@FlorianDecker That works, thanks.

**Comment by FlorianDecker** [comment]
Fixed in [2.0.2](https://github.com/rubicon-oss/LicenseHeaderManager/releases/tag/2.0.2). The update should be released in the Visual Studio Marketplace in the near future.
